Add new `a2a` call task to the specification and schema (#1114)
serverlessworkflow/specification: dsl-reference.md

          status: available
```

##### A2A Call

The [A2A Call](#a2a-call) enables workflows to interact with AI agents described by [A2A](https://a2a-protocol.org/).

###### Properties

| Name | Type | Required | Description|
|:--|:---:|:---:|:---|
| method | `string` | `yes` | The A2A JSON-RPC method to send. |

> | agentCard | [`externalResource`](#external-resource) | `yes` | The AgentCard resource that describes the agent to call. |

Reviewer: I don't think this one should be required. I remember in our discussions @cdavernas mentioning that the vcards are not required between the a2a communication.

Reviewer 2: @ricardozanini it's actually needed, as was pointed out in the weekly, to resolve potential authentication.

Reviewer: But is potential authentication required? I'm inquiring because I'm integrating with a few agents and in dev, it's not. I don't think this field should be required.

Reviewer 2: It shouldn't be required, indeed

Author: I added the server field and made the agentCard not required.

@ricardozanini The `server` field is ignored if the `agentCard` is set. I did this because it matches other parts of the spec but we could clarify it in another way.

Author: @ricardozanini For us so far when we want to have separate environments we have been creating separate Agent Cards so haven't needed an override. Since it's just as easy to change the `agentCard` field as it is to change the `server` field I can't think of any situation where an override would be necessary.

> | parameters | `any` | `no` | The parameters for the A2A rpc method. For the `message/send` and `message/stream` methods the parameters `message.messageId` and `message.role` must be automatically set if missing. |

Reviewer: ```suggestion
| parameters | `any` | `no` | The parameters for the A2A RPC method. For the `message/send` and `message/stream` methods, runtimes must set the parameters `message.messageId` and `message.role` if missing in the definition. |
```

Reviewer: Don't we want parameters in the format `map[string, any]` instead?

Are both valid?

Map:

```yaml
parameters:
     param1: value1
     param2: value2
```

Any:
```yaml
parameters: myparam
```

Author: @ricardozanini Only the first one is valid but I would like to support directly setting parameters from a jq script the same as Set tasks. Which looking at Set task in the spec would make it `map | string`.

> | parameters | `any` | `no` | The parameters for the A2A rpc method. For the `message/send` and `message/stream` methods the parameters `message.messageId` and `message.role` must be automatically set if missing. |

Author: I have changed it to say 

> For the `message/send` and `message/stream` methods, runtimes must default `message.messageId` to a uuid and `message.role` to `user`.

> [!NOTE]

> > The `security` and `securitySchemes` fields of the AgentCard contain authentication requirements and schemes for when communicating with the agent.

Reviewer: I don't see this definition in the table above. The `agentCard` is an external resource, so a file in a remote location? Won't be able to set it in the workflow definition?

Reviewer: Perfect!

> [!NOTE]

> > On success the output is the JSON-RPC result. On failure runtimes must raise an error.

Reviewer: We have error definitions. I believe we should indicate which errors.

Reviewer 2: @emmanuelallen we could put it as status code, because even if it is logically related to http status codes, it is not restricted to them.
I however think your proposal is cleaner. The title could be the code and or a high level description like you proposed, that decision is purely esthetic and should therefore be up to runtimes.

> [!NOTE]
> For `message/stream` and `tasks/resubscribe` methods the output is a sequentially ordered array of all the result objects.

> 

Reviewer: I'd avoid adding three consecutive note boxes like this. We should rewrite as regular paragraphs.
